### PR TITLE
perf: allocator fast path (alternative testing memory size).

### PR DESCRIPTION
**Pull request by crusso**
Experiment with comparing `new_hp` against `memory_size`. Variation of #3090.

- simpler conceptually/less churn
- still rejects failed `memory_grow` (doesn't assume it goes to 65536 pages)
- avoids grow when used heap within physical memory
- no need to pregrow as with xor trick.

I think I am shaving the yak, but let's see what CI reports for perf.

### CODE REVIEW

File: test/bench/alloc.mo [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// The 2 palindrome implementations from
+// "There and Back Again", by Olivier Danvy and Mayer Goldberg
```
Comment (crusso) [inline test/bench/alloc.mo:2]:
```suggestion
```

File: test/bench/alloc.mo [hunk 2]
```diff
@@ -0,0 +1,29 @@
+//MOC-FLAG --force-gc
+import { error; performanceCounter; rts_heap_size; debugPrint ; stableMemoryGrow; stableMemoryLoadBlob} = "mo:⛔";
```
Comment (crusso) [inline test/bench/alloc.mo:2]:
```suggestion
import { error; performanceCounter; rts_heap_size; debugPrint } = "mo:⛔";
```

File: rts/motoko-rts/src/memory/ic.rs [hunk 3]
```diff
@@ -88,17 +85,17 @@ impl Memory for IcMemory {
     }
 }
 
-/// Page allocation. Ensures that the memory up to, and including, the given pointer is allocated,
+/// Page allocation. Ensures that the memory up to, but excluding, the given pointer is allocated,
 /// with the slight exception of not allocating the extra page for address 0xFFFF_0000.
```
Comment (ggreif) [inline rts/motoko-rts/src/memory/ic.rs:89]: with the caveat

File: rts/motoko-rts/src/memory/ic.rs [hunk 4]
```diff
@@ -88,17 +85,17 @@ impl Memory for IcMemory {
...
-/// Enforced precondition: `ptr` must be a pointer less than or equal to 0xFFFF_0000.
 #[inline(never)]
 unsafe fn grow_memory(ptr: u64) {
-    debug_assert_eq!(0xFFFF_0000, usize::MAX - WASM_PAGE_SIZE.as_usize() + 1);
+    debug_assert!(ptr <= 2 * u64::from(core::u32::MAX));
     if ptr > 0xFFFF_0000 {
         // spare the last wasm memory page
-        rts_trap_with("Cannot allocate memory")
+        rts_trap_with("Cannot grow memory")
```
Comment (ggreif) [inline rts/motoko-rts/src/memory/ic.rs]: if you change this you have to re-accept the `test/run-drun-no-ci` tests.

File: rts/motoko-rts/src/memory/ic.rs [hunk 5]
```diff
@@ -88,17 +85,17 @@ impl Memory for IcMemory {
...
 }
 
-/// Page allocation. Ensures that the memory up to, and including, the given pointer is allocated,
+/// Page allocation. Ensures that the memory up to, but excluding, the given pointer is allocated,
 /// with the slight exception of not allocating the extra page for address 0xFFFF_0000.
-/// Enforced precondition: `ptr` must be a pointer less than or equal to 0xFFFF_0000.
 #[inline(never)]
 unsafe fn grow_memory(ptr: u64) {
-    debug_assert_eq!(0xFFFF_0000, usize::MAX - WASM_PAGE_SIZE.as_usize() + 1);
+    debug_assert!(ptr <= 2 * u64::from(core::u32::MAX));
```
Comment (ggreif) [inline rts/motoko-rts/src/memory/ic.rs:92]: This is not needed any more given line 93. Maybe I forgot to remove it from #3090.

File: test/bench/palindrome.mo [hunk 6]
```diff
@@ -1,7 +1,7 @@
 // The 2 palindrome implementations from
 // "There and Back Again", by Olivier Danvy and Mayer Goldberg
 //
-import { error; performanceCounter; rts_heap_size; debugPrint } = "mo:⛔";
+import { error; performanceCounter; rts_heap_size; debugPrint ; stableMemoryGrow; stableMemoryLoadBlob} = "mo:⛔";
```
Comment (ggreif) [inline test/bench/palindrome.mo]:
```suggestion
import { error; performanceCounter; rts_heap_size; debugPrint; stableMemoryGrow; stableMemoryLoadBlob } = "mo:⛔";
```

Comment (ggreif) [reply test/bench/palindrome.mo]: But why is this change needed?

**Review by ggreif — COMMENTED** [review]
Added comments.

File: rts/motoko-rts/src/memory/ic.rs (hunk 3 above)
Comment (crusso) [reply rts/motoko-rts/src/memory/ic.rs:89]: Yep! Well spotted!

File: rts/motoko-rts/src/memory/ic.rs (hunk 4 above)
Comment (crusso) [reply rts/motoko-rts/src/memory/ic.rs]: Oh, I though you had changed it to this message.

File: rts/motoko-rts/src/memory/ic.rs (hunk 5 above)
Comment (crusso) [reply rts/motoko-rts/src/memory/ic.rs:92]: Done

File: test/bench/palindrome.mo (hunk 6 above)
Comment (crusso) [reply test/bench/palindrome.mo]: It's not there, is it? Weird.

**Review by ggreif — APPROVED** [review]
Cool, this is a nice improvement!

**Pull request merged**